---
title: "Drugs NAFLD"
author: "Maria Jimenez-Ramos"
date: "05/07/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 7, fig.fullwidth =TRUE)
```

We are going to try to replicate the DrugVsDisease package in order to use the drug networks to look for gene signatures. With our distance matrix, we are going to create a function that gives back the different gene signatures for each drug.  

```{r}
PRL_genes <- read.csv("./Data/PRLs_ranked_woBRD.csv", row.names=1) #Upload file with ranked lists of each drug (once summarised across all conditions) from CMap
PRL_genes <- as.matrix(PRL_genes)
AP_communities <- read.delim("./Data/AP_communities_woBRD.txt", sep=  " ") #Upload file of drug network obtained with PRLs file
colnames(AP_communities) <- c("Cluster", "Drug")
geneinfo_matrix <- data.table::fread('D:/CMap/Data/geneinfo_beta.txt')
Cirrhosis_signature <- read.csv("E:/Maria/CMap/Signatures/Cirrhosis.csv") #Disease signature 
Cirrhosis_signature <- Cirrhosis_signature[which(Cirrhosis_signature$adj.P.Val < 0.05 & Cirrhosis_signature$P.Value < 0.05),]
Cirrhosis_signature <- Cirrhosis_signature[order(Cirrhosis_signature$logFC, decreasing = TRUE),]
genes_list <- geneinfo_matrix[match(Cirrhosis_signature$genes, geneinfo_matrix$gene_symbol),] #Match gene symbols to gene ids
genes_list$logFc <- Cirrhosis_signature$logFC 
genes_list <- genes_list[which(!is.na(genes_list$gene_id)),]
test <- as.matrix(genes_list$logFc)
rownames(test) <- genes_list$gene_id #Final matrix with disease signature to use
```

We create a set of functions that are going to be helpful to obtain the final drugs according to the disease profile.

```{r}
#Function to rank a matrix. It takes as input the matrix to be ranked, the dimension where the ranking is going to be applied to in the matrix and boolean value -TRUE or FALSE- indicating whether the matrix should be ranked in decreasing or ascending order.
rankmat <- function(data, ref, decreasing){
  if(is.matrix(data)){
    rankmatrix <- apply(data, ref, function(x){names(x)[sort.list(as.integer(x), decreasing = decreasing, method = "radix")]})
  }else{
    rankedmat <- names(sort(data,decreasing = decreasing))
  }
 return(rankmatrix) 
}

#Function to perform the Kolmogorov-Smirnov statistics that is going to take as input a ranked matrix obtained from random profiles for permutation tests, the reference ranked matrix that is going to be compared to and the length of the profile.
KS <- function(refmatrix, inputmatrix, lengthtest = 250){
  refmatrix <- as.matrix(refmatrix)
  inputmatrix <- as.matrix(inputmatrix)
  N <- lengthtest
  Nh <- nrow(refmatrix)
  nototest <- ncol(inputmatrix)
  outmat <- c()
  for (i in 1:nototest){
    toptail <- inputmatrix[1:lengthtest, i]
    toptoref <- refmatrix%in%toptail
    In <- matrix(0, nrow = nrow(refmatrix), ncol = ncol(refmatrix))
    Out <- matrix(0, nrow = nrow(refmatrix), ncol = ncol(refmatrix))
    In[toptoref] <- 1/N
    Out[!toptoref] <- 1/(Nh-N)
    if(ncol(refmatrix)==1){
      In <- as.matrix(In, ncol = 1)
      Out <- as.matrix(Out, ncol = 1)
    }
    In <- apply(In, 2, cumsum)
    Out <- apply(Out, 2, cumsum)
    dif <- In-Out
    refvalues <- apply(dif, 2, function(x){x[which.max(abs(x))]})
    outmat <- rbind(outmat, refvalues)
  }
  rownames(outmat) <- colnames(inputmatrix)
  colnames(outmat) <- colnames(refmatrix)
  return(outmat)
}

#This funtion is going to perform a weight signed ranked score. The input is the same as the function before.
WSR <- function(refmatrix, inputmatrix, uplength = 250, downlength = 250){
  refmatrix <- as.matrix(refmatrix)
  inputmatrix <- as.matrix(inputmatrix)
  N <- nrow(refmatrix)
  m <- uplength+downlength
  inputranksignmat <- apply(inputmatrix,2, function(x){rank(abs(x))})
  refranksignmat <- apply(refmatrix, 2, function(x){rank(abs(x))})
  name_input <- rankmat(inputranksignmat,2, TRUE)
  geneset <- as.matrix(name_input[1:m,])
  stat <- matrix(nrow = ncol(inputmatrix), ncol = ncol(refmatrix))
  for (i in 1:ncol(inputmatrix)){
    inputrank <- rank(abs(inputmatrix[geneset[,i],i]))
    inputvals <- inputrank*sign(inputmatrix[geneset[,i],i])
    refvals <- refranksignmat[geneset[,i],]*sign(refmatrix[geneset[,i],])
    statvec <- refvals*inputvals
    stat[i,] <- apply(statvec, 2, sum)
  }
  max_score <- sum((N-1:m+1)*(m-1:m+1))
  stat <- stat/max_score
  return(stat*-1)
}

#This function will provide the ES scores.
getES_scores <- function(diseasematup, diseasematdown, uplength = 250, downlength = 250, PRLs, stat = c("KS", "WSR")){
  if(stat=="KS"){
    PRLs <- abs(PRLs)
    ranknamePRLs <- rankmat(PRLs, 2, TRUE)
    ranknamePRLs_inverse <- rankmat(PRLs, 2, FALSE)
    up_profile <- KS(refmatrix = ranknamePRLs, inputmatrix = diseasematup, lengthtest = uplength)
    up_ref <- KS(refmatrix = diseasematup, inputmatrix= ranknamePRLs, lengthtest= uplength)
    up_ref <- t(up_ref)
    down_profile <- KS(refmatrix = ranknamePRLs, inputmatrix = diseasematdown, lengthtest = downlength)
    down_ref <- KS(refmatrix = diseasematup, inputmatrix = ranknamePRLs_inverse, lengthtest = downlength)
    down_ref <- t(down_ref)
    
    score1 <- (up_profile - down_profile)/2
    score2 <- (up_ref - down_ref)/2
    check1 <- sign(up_profile) == sign(down_profile)
    check2 <- sign(up_ref) == sign(down_ref)
    score1[check1] <- 0
    score2[check2] <- 0
    scores <- score1+score2
    all <- as.matrix(scores/2)
  }else{
    all <- WSR(refmatrix = PRLs, inputmatrix = diseasematup, uplength = uplength, downlength = downlength)
    all <- as.matrix(all)
  }
  if (is.matrix(diseasematup)){
    rownames(all) <- colnames(diseasematup)
  }
  colnames(all) <- colnames(PRLs)
  return(all)
}

findSignifCompounds <- function(scores, rankdata = NULL, rankdata_inverse = NULL, uplength = 250, downlength = 250, stat, nperm = 100, signif.fdr){
  rankdata <- as.matrix(rankdata)
  size <- nrow(rankdata)
  smatrix <- sapply(1:nperm, function(x){sample.int(size)})
  rownames(smatrix) <- rankdata[,1]
  samplemat <- rankmat(smatrix, 2, TRUE)
  samplemat_inverse <- rankmat(smatrix, 2, FALSE)
  if (stat == "KS"){
    permup <- KS(refmatrix = samplemat, inputmatrix = rankdata, lengthtest = uplength)
    permupref <- KS(refmatrix = rankdata, inputmatrix = samplemat, lengthtest = uplength)
    permupref <- t(permupref)
    permdown <- KS(refmatrix = samplemat, inputmatrix = rankdata_inverse, lengthtest = downlength)
    permdownref <- KS(refmatrix = rankdata, inputmatrix = samplemat_inverse, lengthtest = downlength)
    permdownref <- t(permdownref)
    score1 <- (permup-permdown)/2
    score2 <- (permupref-permdownref)/2
    check1 <- sign(permup) == sign(permdown)
    check2 <- sign(permupref) == sign(permdownref)
    score1[check1] <- 0
    score2[check2] <- 0
    Pscores <- score1+score2
    Pscores <- Pscores/2
  }else{
    samplesign <- sapply(1:nperm, function(x){sample(c(-1,1), size = size, replace =TRUE)})
    samplemat <- smatrix*samplesign
    rownames(samplemat) <- rownames(rankdata)
    Pscores <- WSR(refmatrix = samplemat, inputmatrix = rankdata, uplength = uplength, downlength = downlength)
  }
  empcdf <- ecdf(abs(Pscores))
  emppvals <- vector(length=length(scores))
  emppvals <- 1-empcdf(abs(scores))
  pvaladj <- p.adjust(emppvals, "BH")
  sel <- which(pvaladj < signif.fdr) #SI CAMBIO ESTO PUEDO CONSEGUIR TODOS LOS COMPUESTOS INDEPENDIENTEMENTE DEL PVAL?????
  signifcompounds <- scores[sel]
  names(signifcompounds) <- names(scores)[sel]
  return(signifcompounds)
}

#This function gets permutation based empirical p-values for enrichment scores.	
significantES <- function(ESscores, rankdata = NULL, rankdata_inverse = NULL, uplength = 250, downlength = 250, stat, nperm, signif.fdr){
  if (is.list(ESscores)){
    results <- list(length = length(ESscores))
    for (i in 1:length(ESscores)){
    scorematrix <- ESscores[[i]]
    if(is.matrix(scorematrix)){
      signifcompounds <- list(length = nrow(scorematrix))
      for (i in 1:nrow(scorematrix)){
        signifcompounds[[i]] <- findSignifCompounds(scorematrix[i,], rankdata = rankdata[,i], rankdata_inverse = rankdata_inverse[,i], uplength = uplength, downlength = downlength, stat = stat, nperm = nperm, signif.fdr = signif.fdr)
      }
      names(signifcompounds) <- rownames(scorematrix)
    }else{
      signifcompounds <- findSignifCompounds(scorematrix, rankdata = rankdata, rankdata_inverse = rankdata_inverse, uplength = uplength, downlength =downlength, stat = stat, nperm = nperm, signif.fdr = signif.fdr)
    }
    results[[i]] <- signifcompounds
    }
    names(results) <- names(ESscores)
  }else{
   scorematrix <- ESscores
   if (is.matrix(scorematrix)){
     signifcompounds <- list(length = nrow(scorematrix))
     for (i in 1:nrow(scorematrix)){
       signifcompounds[[i]] <- findSignifCompounds(scorematrix[i,], rankdata = rankdata[,i], rankdata_inverse = rankdata_inverse[,i], uplength = uplength, downlength = downlength, nperm = nperm, stat = stat, signif.fdr = signif.fdr)
     }
     names(signifcompounds) <- rownames(scorematrix)
    }else{
      signifcompounds <- list(length = 1)
      signifcompounds[[1]] <- findSignifCompounds(scorematrix, rankdata = rankdata, rankdata_inverse = rankdata_inverse, uplength = uplength, downlength = downlength, stat= stat, signif.fdr = signif.fdr)
      names(signifcompounds) <- names(ESscores)
    }
    results <- signifcompounds
  }
  return(results)
}
```

We create another function to calculate the ES score from an input profile. The input is the following:

*`data`: matrix gene expression profiles, where the rows are the genes and columns are different profiles.
*`lengthtest`: integer giving the number of genes that will be used to generate the set size to use for look for drug profiles.
*`signif.fdr`: false discovery rate that will be used to determine the significance of enrichment scores.
*`nperm`: integer for the number of permutation profiles.
*`PRLs`: the PRLs lists that contain the drug profiles  -not ranked, not distance.
*`stat`: either Kolmogorov-Smirnov (KS) or Weigthed Signed Ranked Score (WSR). The first one equally weights all elements of the gene set, while the second one uses both the sign and the position in the ranked list to calculate the enrichment scores.

The output is a list with the ES scores and their significance.

```{r}
calculateES <- function(data, PRLs, stat=c("WSR", "KS"), uplength = uplength, downlength =downlength, signif.fdr, nperm, pvalues = NULL, type= c("fixed", "dynamic"), dynamic.fdr = 0.05){
  if(stat == "WSR"){
    ngenes <- nrow(PRLs)
    splitn <- round(ngenes/2)+1
    refDB <- splitn - PRLs
  }else{
    refDB <- PRLs
  }
  refDB <- as.matrix(refDB)
  data <- as.matrix(data)
  lt<-sum(rownames(data)%in%rownames(refDB))
        if (lt!=nrow(data)){
          stop('Rownames of input data and reference data do not match')}
  refDB <- refDB[rownames(data),]
  if(stat=="KS"){
    datarl <- apply(data, 2, function(x){as.integer(rank(x))})
    datarl <- as.matrix(datarl)
    rownames(datarl) <- rownames(data)
    colnames(datarl) <- colnames(data)
    rankdata <- rankmat(datarl, 2, TRUE)
    rankdata_inverse <- rankmat(datarl, 2, FALSE)
  }else{
    rankdata <- data
    rankdata_inverse <- data
  }
  if (type == "fixed"){
    outk <- getES_scores(rankdata, rankdata_inverse, PRLs = refDB, uplength = uplength, downlength = downlength, stat = stat)
    signif <- significantES(ESscores = outk, rankdata = rankdata, rankdata_inverse = rankdata_inverse, signif.fdr = signif.fdr, uplength = uplength, downlength = downlength, stat = stat, nperm= nperm)
  names(signif) <- colnames(data)
  }else{
    pvals <- as.matrix(pvalues)
    pvals <- apply(pvals, 2, p.adjust)
    signifgenes <- apply(pvals, 2, function(x){x<dynamic.fdr})
    rankdrugPRLs <- rankmat(refDB, 2, TRUE)
    rankdrugPRLs_inverse <- rankmat(refDB, 2, FALSE)
    if (ncol(pvals)>1){
      signif <- list(length = ncol(pvals))
      outk <- list(length = ncol(pvals))
      postmat <- apply(data,2,function(x){x>0})
      upsignif <- postmat*signifgenes
      for (i in 1:ncol(pvals)){
        if (sum(signifgenes[,i])>0){
          uplength <- sum(upsignif[,i])
          downlength <- sum(signifgenes[,i])-uplength
          if (uplength <50){
            warning("Number of significantly up-regulated genes is less than 50, top 50 used instead")
            uplength<-50
          }
          if (downlength <50){
            warning("Number of significantly down-regulated genes is less than 50, top 50 used instead")
            downlength<-50
          }
          outk[[i]] <- getES_scores(rankdata[,i], rankdata_inverse[,i], uplength = uplength, downlength = downlength, PRLs= refDB, stat = stat)
          signif[[i]] <- significantES(ESscores = outk[[i]], uplength = uplength, downlength= downlength, rankdata = rankdata, rankdata_inverse = rankdata_inverse, nperm =nperm, signif.fdr = signif.fdr, stat = stat)
        }else{
          warning(paste("No significantly differentially expressed genes for experiment",colnames(pvalues)[i]))
          outk[[i]]<-NULL
          signif[[i]]<-NULL
        }
      }
      names(outk) <- colnames(data)
      names(signif) <- colnames(data)
    }else{
       if (sum(signifgenes)>0){
         postmat <- data >0
         upsignif <- postmat*signifgenes
         uplength <- sum(upsignif)
         downlegnth <- sum(signifgenes) - uplength
         if (uplength <50){
           warning("Number of significantly up-regulated genes is less than 50, top 50 used instead")
            uplength<-50
         }
         if (downlength < 50){
           warning("Number of significantly down-regulated genes is less than 50, top 50 used instead")
            downlength<-50
         }
         outk <- getESscores(rankdata, rankdata_inverse, PRLs = refDB, uplength = uplength, downlength = downlength, stat = stat)
         signif <- signficantES(ESscores = ouk, rankdata= rankdata, rankdata_inverse = rankdata_inverse, nperm = nperm, signif.fdr = signif.fdr, uplength = uplength, downlength =downlength, stat = stat)
       }else{
         stop(paste("No significantly differentially expressed genes, consider changing FDR threshold"))
       }
      names(signif) <- colnames(data)
      names(outk) <- colnames(data)
    }
  }
  return(list(scores = outk, significance = signif))
}
```

The final drugs will be clustered together by an average linkage cluster following DrugVsDisease package. Two functions have been used to to achieve this. The input is the following:

*`ESscores`: enrichment scores obtaind by comparing the input profile to the drug network.
*`statistics`: whether the median or the mean is used to perform the average cluster. 
*`drugClusters`: data frame that contains the different drugs and the cluster where it belongs.
*`stat`: whether WSR or KS is used to get the enrichment scores.

`classifyaverageprofile` output is a data frame that will contained the drugs with the ES scores, the cluster and the RPS.

```{r}
averagecluster <-
function(ESscores, statistic = c("mean", "median"), drugClusters, stat){
	#get the number of clusters available
	ncluster <- unique(drugClusters[,"Cluster"])
	numberclusters <-length(ncluster)
	avgscores <- rep(0,numberclusters)
	medscores <- rep(0,numberclusters)
	avgscoresneg <- rep(0,numberclusters)
	medscoresneg <- rep(0,numberclusters)
	scorepos <- ESscores > 0
	#for each cluster calculate the mean and median scores
	for(k in 1:numberclusters){
			selc <- drugClusters[which(drugClusters[,"Cluster"]==ncluster[k]),"Drug"]
			p <- colnames(scorepos)[scorepos]
			n <- colnames(scorepos)[!scorepos]
			sp <- selc[selc%in%p]
			sn <- selc[selc%in%n]
		if(length(sp)>0){	
			avgscores[k]<-mean(ESscores[1,][sp])		
			medscores[k]<-median(ESscores[1,][sp])
		}
		if(length(sn)>0){
			avgscoresneg[k]<-mean(abs(ESscores[1,][sn]))
			medscoresneg[k]<-median(abs(ESscores[1,][sn]))				
		}	
	}
	#now assign to the cluster with the maximum score:
	if(statistic=="mean"){
	  cluster <- which.max(avgscores)
	  clusterneg <- which.max(avgscoresneg)
	}else{
	  cluster <- which.max(medscores)
	  clusterneg <- which.max(medscoresneg)
	}
	
	#now return the scores to elements in the max cluster, sel has the names of the drugs
	sel <- drugClusters[which(drugClusters[,"Cluster"]==ncluster[cluster]),"Drug"]
  selneg <- drugClusters[which(drugClusters[,"Cluster"]==ncluster[clusterneg]),"Drug"]	
	selscores <- ESscores[1,][as.character(sel)]
	selscoresneg <- ESscores[1,][as.character(selneg)]
	distscores <- 1-abs(selscores)
	distscoresneg <- 1-abs(selscoresneg)
	
	if(stat=="KS"){
	results <- data.frame(as.character(sel),distscores,rep(ncluster[cluster],length(sel)),rep(-1,length(sel)))
	resultsneg <- data.frame(as.character(selneg),distscoresneg,rep(ncluster[clusterneg],length(selneg)),rep(1,length(selneg)))
	}else{
	results <- data.frame(as.character(sel),distscores,rep(ncluster[cluster],length(sel)),rep(1,length(sel)))
	resultsneg <- data.frame(as.character(selneg),distscoresneg,rep(ncluster[clusterneg],length(selneg)),rep(-1,length(selneg)))
		}
	colnames(results) <- c("Drugs","ES Distance","Cluster","RPS")
	colnames(resultsneg) <- c("Drugs","ES Distance","Cluster","RPS")
	return(rbind(results,resultsneg))
}

classifyaveragelinkage <-
function(ESscores,statistic=c("mean","median"), stat, drugClusters){
  if (is.list(ESscores)){
    ClusterAssignments <- list(length = length(ESscores))
    for (i in 1:length(ESscores)){
      scorematrix <- ESscores[[i]]
      if(is.matrix(scorematrix)){
        avgscores <- list(length=nrow(scorematrix))
        for (j in 1:nrow(scorematrix)){
          avgscores[[j]] <- averagecluster(ESscores = scorematrix[j,], statistic = statistic, drugClusters = drugClusters, stat = stat)
        }
        names(avgscores) <- rownames(scorematrix)
      }else{
        avgscores <- averagecluster(ESscores = scorematrix, statistic = statistic, drugClusters = drugClusters, stat=stat)
      }
      ClusterAssignments[[i]] <- avgscores
    }
    names(ClusterAssignments) <- names(ESscores)
  }else{
    scorematrix <- ESscores
		if(is.matrix(scorematrix)){
			avgscores <-list(length=nrow(scorematrix))				
				for(j in 1:nrow(scorematrix)){
					avgscores[[j]]<- averagecluster(ESscores = scorematrix[j,],statistic = statistic, drugClusters=drugClusters, stat = stat)
				}
			names(avgscores) <-rownames(scorematrix)				
		}else{				
			avgscores <-list(length=1)
			avgscores <-averagecluster(ESscores = scorematrix,statistic=statistic,drugClusters=drugClusters, stat= stat)
			names(avgscores) <- names(scorematrix)
		}
    ClusterAssignments <- avgscores
  }
	return(ClusterAssignments)
}
```

Single clustering

```{r}
findCluster <-
function(SignifScores, drugClusters=NULL, no.signif=30){
	#check that we have some significant results otherwise return na
	if (length(SignifScores)==0){return(NA)}
	if (is.character(SignifScores)){return(NA)}
	#check if there are more significant results than have been asked to assign to a cluster
	if(length(SignifScores) > no.signif){cat(paste('Number of Significant results greater than', no.signif, 'Using top',no.signif, 'hits - consider using average linkage instead'),fill=TRUE)}
	#order by most significant results (i.e. the highest absolute Enrichment scores) and take the top no.signif to classify
	absscores <- abs(SignifScores)
	SortScores <- sort(absscores,TRUE)
	SScores <- SignifScores[names(SortScores)]
	SignifScores <- SScores[1:no.signif]
	signifnames <-names(SignifScores)
	#find the cluster for the significant matches, and output to data frame along with the distance between profiles (1-ES), and name of the profile.
	sel <- sapply(signifnames,function(x){which(drugClusters[,"Drug"]==x)})
	sel <- unlist(sel)
	scores <- SignifScores[names(sel)]
	distscores<-1-abs(scores)
	cors<-ifelse(scores>0,-1,1)
	results.clust<- data.frame(names(sel), distscores, drugClusters[sel,"Cluster"], cors)
	colnames(results.clust)<-c("Drug","ES Distance","Cluster","RPS")
return(results.clust)
}

classifysinglelinkage <-
function(significance,drugClusters=NULL,no.signif=30){
	#significance output from significantES
	#if length(significance)>1 then we have dynamic set sizes for the KS scor
	Clusterlist<-list(length=length(significance))
		for(i in 1:length(significance)){		
			SignifScores <- significance[[i]]
			#see if SignifScores is a list - i.e. more than one disease profile
			if(is.list(SignifScores)){
				Clust<-lapply(SignifScores, function(x){findCluster(x, drugClusters = drugClusters, no.signif = 10)})
				names(Clust) <- names(SignifScores)
			}else{
				#just have one profile:
				Clust<- findCluster(SignifScores,drugClusters= drugClusters,no.signif=no.signif)
			}
			Clusterlist[[i]] <- Clust		
		}
		names(Clusterlist) <- names(significance)
	return(Clusterlist)
}

```


We can finally now create the final function that will give us the drug profiles that could be used as treatment for our ranked gene profile. The input of the function is:

*`data`: matrix gene expression profiles, where the rows are the genes and columns are different profiles.
*`lengthtest`: integer giving the number of genes that will be used to generate the set size to use for look for drug profiles.
*`signif.fdr`: false discovery rate that will be used to determine the significance of enrichment scores.
*`nperm`: integer for the number of permutation profiles.
*`avgstat`: character string that indicates wheter the median or the mean is used to cluster the drugs in the average linked cluster.
*`no.signif`: integer giving the number of significant enrichment scores to return. Default is 10.
*`stat`: either Kolmogorov-Smirnov (KS) or Weigthed Signed Ranked Score (WSR). The first one equally weights all elements of the gene set, while the second one uses both the sign and the position in the ranked list to calculate the enrichment scores.

The output is a data frame that will contain the names of the profiles with significant scores, the distance between the input profile and the reference profile (1- ES), the cluster number of the node.


```{r}
classifyprofile <- function(data, pvalues = NULL, PRLs, type = c("fixed", "dynamic"), uplength = 100, downlength = 100, signif.fdr=0.05, dynamic.fdr = 0.05, nperm=1000, avgstat=c("mean","median"), no.signif=30, stat=c("KS","WSR"), drugClusters, cluster = c("average", "single")){
	if(!is.matrix(data)){
		data<-read.table(data,header=TRUE,row.names=1)
		data<-as.matrix(data)
	}
  if(!is.null(pvalues)&&!is.matrix(pvalues)){
    pvalues<-read.table(pvalues,header=TRUE,row.names=1)
    pvalues<-as.matrix(pvalues)
  }
  if(!is.matrix(PRLs)){PRLs<-read.table(PRLs,header=TRUE,row.names=1)}
  if(!is.data.frame(drugClusters)){drugClusters<-read.table(drugClusters,header=TRUE)}
  refnames<-colnames(PRLs)
  refnames<-make.names(refnames)
  clustnames<-drugClusters[,which(colnames(drugClusters)%in%c("Drug","Disease"))]
  clustnames<-make.names(clustnames)
  intersection<-which(refnames%in%clustnames)
  if(length(refnames)!=length(intersection)){stop('Profiles in custom rank profiles and clustom clusters do not match')}
	#calculate enrichment scores
	ESvals<- calculateES(data,PRLs,type = type, dynamic.fdr = dynamic.fdr, signif.fdr=signif.fdr,nperm=nperm,uplength=uplength, downlength = downlength,stat=stat)
	#assign significant enrichment scores to clusters	
	if (cluster == "single"){
	  clusterassignments<- classifysinglelinkage(ESvals$significance, drugClusters= drugClusters,no.signif)
	}else{
	  clusterassignments<- classifyaveragelinkage(ESvals$scores, statistic= avgstat, drugClusters = drugClusters, stat = stat)
	}
	#check have at least one significant results
	nacheck<-unlist(clusterassignments)
	if(is.list(nacheck)){
		nacheck2<-unlist(nacheck)
		nas<-which(is.na(nacheck2))
		if(length(nas)==length(nacheck2)){warning('No significant matches found. Consider changing the FDR threshold')}
	}else{
		nas<-which(is.na(nacheck))
		if(length(nas)==length(nacheck)){warning('No significant matches found. Consider changing the FDR threshold')}
	}
	return(clusterassignments)
}
```

Next, we upload all the disease signatures and obtain the drugs according to both methods. The ones present in both methods will be the final drugs used. 
```{r}

steatosis <- read.csv("/Users/mariajimenezramos/Library/CloudStorage/OneDrive-UniversityofEdinburgh/PhD/Third year/SteatoSITE/Results_project/Without non-prot coding genes/DEG/Simple_steatosis.csv")
f0f1 <- read.csv("/Users/mariajimenezramos/Library/CloudStorage/OneDrive-UniversityofEdinburgh/PhD/Third year/SteatoSITE/Results_project/Without non-prot coding genes/DEG/F0F1.csv")
f2 <- read.csv("/Users/mariajimenezramos/Library/CloudStorage/OneDrive-UniversityofEdinburgh/PhD/Third year/SteatoSITE/Results_project/Without non-prot coding genes/DEG/F2.csv")
f3 <- read.csv("/Users/mariajimenezramos/Library/CloudStorage/OneDrive-UniversityofEdinburgh/PhD/Third year/SteatoSITE/Results_project/Without non-prot coding genes/DEG/F3.csv")
f4 <- read.csv("/Users/mariajimenezramos/Library/CloudStorage/OneDrive-UniversityofEdinburgh/PhD/Third year/SteatoSITE/Results_project/Without non-prot coding genes/DEG/F4.csv")


test <- list()
signatures <- list(steatosis, f0f1, f2, f3, f4)

for (i in signatures){
  i <- i[which(abs(i$logFC) >= 1 & i$adj.P.Val < 0.05),]
  i <- i[order(i$logFC, decreasing = TRUE),]
}

for (i in 1:5){
  genes_list <- geneinfo_matrix[match(signatures[[i]]$genes, geneinfo_matrix$gene_symbol),]
  genes_list$logFc <- signatures[[i]]$logFC 
  genes_list <- genes_list[which(!is.na(genes_list$gene_id)),]
  test[[i]] <- as.matrix(genes_list$logFc)
  rownames(test[[i]]) <- genes_list$gene_id
}

#Obtain drugs with the KS signature
steatosis_drugs_ks <- classifyprofile(test[[1]], PRLs = PRLs, type = "fixed", uplength = 44, downlength = 100, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "KS", drugClusters = ap_communities, cluster = "single")
f0f1_drugs_ks <- classifyprofile(test[[2]], PRLs = PRLs, type = "fixed", uplength = 56, downlength = 80, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "KS", drugClusters = ap_communities, cluster = "single")
f2_drugs_ks <- classifyprofile(test[[3]], PRLs = PRLs, type = "fixed", uplength = 100, downlength = 91, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "KS", drugClusters = ap_communities, cluster = "single")
f3_drugs_ks <- classifyprofile(test[[4]], PRLs = PRLs, type = "fixed", uplength = 100, downlength = 100, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "KS", drugClusters = ap_communities, cluster = "single")
f4_drugs_ks <- classifyprofile(test[[5]], PRLs = PRLs, type = "fixed", uplength = 100, downlength = 100, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "KS", drugClusters = ap_communities, cluster = "single")

#Obtain drugs with the WSR signature
steatosis_drugs_wsr <- classifyprofile(test[[1]], PRLs = PRLs, type = "fixed", uplength = 44, downlength = 100, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "WSR", drugClusters = ap_communities, cluster = "single")
f0f1_drugs_wsr <- classifyprofile(test[[2]], PRLs = PRLs, type = "fixed", uplength = 56, downlength = 80, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "WSR", drugClusters = ap_communities, cluster = "single")
f2_drugs_wsr <- classifyprofile(test[[3]], PRLs = PRLs, type = "fixed", uplength = 100, downlength = 91, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "WSR", drugClusters = ap_communities, cluster = "single")
f3_drugs_wsr <- classifyprofile(test[[4]], PRLs = PRLs, type = "fixed", uplength = 100, downlength = 100, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "WSR", drugClusters = ap_communities, cluster = "single")
f4_drugs_wsr <- classifyprofile(test[[5]], PRLs = PRLs, type = "fixed", uplength = 100, downlength = 100, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "WSR", drugClusters = ap_communities, cluster = "single")

#Intersect between the drug signatures
steatosis_drugs_finals <- intersect(steatosis_drugs_ks[[1]]$Drug, steatosis_drugs_wsr[[1]]$Drug)
f0f1_drugs_finals <- intersect(f0f1_drugs_ks[[1]]$Drug, f0f1_drugs_wsr[[1]]$Drug)
f2_drugs_finals <- intersect(f2_drugs_ks[[1]]$Drug, f2_drugs_wsr[[1]]$Drug)
f3_drugs_finals <- intersect(f3_drugs_ks[[1]]$Drug, f3_drugs_wsr[[1]]$Drug)
f4_drugs_finals <- intersect(f4_drugs_ks[[1]]$Drug, f4_drugs_wsr[[1]]$Drug)
```

To make sure that the compounds I am obtaining are not false positives, I can also create fake gene signatures to observe if these compounds would also show up in here. I will create 100 different gene signatures and observe which are the most common compounds that I obtain when creating them. If there are some compounds that show most than the rest it would be a sign that they are prone to show up more than the rest and could be potential false positives. 

I am going to create a matrix that will contain 2000 genes and 100 samples. Then I will randomly choose 800 of those genes for further analyses.

```{r}
# Set the number of signatures and genes
num_signatures <- 100
num_genes <- 800

# Load a sample gene list (you can replace this with your own gene list)
library(biomaRt)

# Use the Ensembl database to retrieve gene names for humans
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
genes <- getBM(attributes = c("hgnc_symbol"), mart = ensembl)

# Extract the gene names from the gene list
gene_names <- genes$hgnc_symbol

# Create a list to store the signatures
signatures <- list()

# Generate the signatures
set.seed(123)
for (i in 1:num_signatures) {
  # Sample 800 genes from the gene list without replacement
  signature_genes <- sample(gene_names, num_genes, replace=FALSE)
  # Add the signature genes to the signatures list
  signatures[[i]] <- signature_genes
}

drugs <- list()

for (i in 1:length(signatures)){
  genes_list <- geneinfo_matrix[match(signatures[[i]], geneinfo_matrix$gene_symbol),]
  genes_list$logFc <-  nrow(genes_list):1
  genes_list <- genes_list[which(!is.na(genes_list$gene_id)),]
  signature <- as.matrix(genes_list$logFc)
  rownames(signature) <- genes_list$gene_id
  drug <- classifyprofile(signature,PRLs = PRLs, type = "fixed", uplength = 100, downlength = 100, signif.fdr = 0.05, nperm = 1000, avgstat = "mean", no.signif = 30, stat = "KS", drugClusters = ap_communities, cluster = "single")
  if (is.null(drug)) {
    drugs[[i]] <- NA
  } else {
    drugs[[i]] <- as.data.frame(drug[[1]])
  }
}

# extract drugs of each dataframe and store as a list
col_list <- lapply(drugs, function(df) df[[1]][1])

# create a named list of unique values
unique_vals <- list()
for (i in 1:length(col_list)) {
  unique_vals[[i]] <- unique(col_list[[i]])
  names(unique_vals)[i] <- paste0("df", i)
}

# create a frequency table of unique values
freq_table <- as.data.frame(table(unlist(unique_vals)))

# plot the frequency table
barplot(freq_table, main = "Frequency of unique values in first column of dataframes")
ggplot(freq_table, aes(x = Var1, y = Freq)) + 
  geom_bar(stat = "identity") + 
  xlab("Drugs") + 
  ylab("Frequency") + 
  theme(axis.text.x = element_blank())
```

